File: BookShelf/ObjC/Cell/BookAutocompleteCell.m
//
//  BookAutocompleteCell.m
//  Bookshelf
//
//  Created by Den Jo on 2020/10/19.
//

#import "BookAutocompleteCell.h"
#import "Bookshelf-Swift.h"

@interface BookAutocompleteCell()

// MARK: - IBOutlet
@property (strong, nonatomic) IBOutlet UIImageView *bookImageView;
@property (strong, nonatomic) IBOutlet UILabel *titleLabel;
@property (strong, nonatomic) IBOutlet UILabel *subtitleLabel;
@property (strong, nonatomic) IBOutlet UILabel *isbnLabel;
@property (strong, nonatomic) IBOutlet UILabel *price;
@property (strong, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicatorView;


// MARK: - Value
@property (strong, nonatomic, nullable) NSURL *imageURLCache;

@end


@implementation BookAutocompleteCell

// MARK: - View Life Cycle
- (void)awakeFromNib {
    [super awakeFromNib];
    
    self.imageView.image = nil;
    [[ImageDataManager shared] cancelDownloadWithUrl: self.imageURLCache];
}



// MARK: - Function
- (void)update: (BookAutocomplete *) data {
    // Image
    self.imageURLCache = data.imageURL;  // Cache

    [self.activityIndicatorView startAnimating];
    [[ImageDataManager shared] downloadWithUrl: data.imageURL completion:^(NSURL * _Nullable url, UIImage * _Nullable image) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if (self == nil) {
                return;
            }
        
            [self.activityIndicatorView stopAnimating];
            
            if (url == nil) {
                return;
            }
            
            if (self.imageURLCache.hash == url.hash) {
                self.imageView.image = image;
                
            } else {
                self.imageView.image = nil;
            }
        });
    }];
    
    
    // Info
    self.titleLabel.text    = data.title;
    self.subtitleLabel.text = data.subtitle;
    self.isbnLabel.text     = data.isbn;
    self.price.text         = data.price;
}

@end

File: BookShelf/Search/Cell/BookAutocompleteCell.swift
//
//  BookAutocompleteCell.swift
//  BookShelf
//
//  Created by Den Jo on 2020/10/20.
//

import UIKit

final class BookAutocompleteCell: UITableViewCell {

    // MARK: - IBOutlet
    @IBOutlet private var bookImageView: UIImageView!
    @IBOutlet private var activityIndicatorView: UIActivityIndicatorView!
    @IBOutlet private var titleLabel: UILabel!
    @IBOutlet private var subtitleLabel: UILabel!
    @IBOutlet private var isbnLabel: UILabel!
    @IBOutlet private var priceLabel: UILabel!
    
    
    
    // MARK: - Value
    // MARK: Public
    static let identifier = "BookAutocompleteCell"
    
    // MARK: Private
    private var imageURL: ImageURL? = nil
    
    
    
    // MARK: - View Life Cycle
    override func prepareForReuse() {
        super.prepareForReuse()
        
        bookImageView.image = nil
        ImageDataManager.shared.cancelDownload(url: imageURL)
    }
    

    
    // MARK: - Function
    // MARK: Public
    func update(data: BookAutocomplete) {
        // Image
        imageURL = ImageURL(url: data.imageURL, hash: hash)
        activityIndicatorView.startAnimating()
        
        ImageDataManager.shared.download(url: imageURL) { [weak self] (url, image) in
            DispatchQueue.main.async {
                guard self?.imageURL == url else { return }
                self?.activityIndicatorView.stopAnimating()
                self?.bookImageView.image = image
            }
        }
        
        
        // Labels
        titleLabel.text    = data.title
        subtitleLabel.text = data.subtitle
        isbnLabel.text     = "ISBN: \(data.isbn)"
        priceLabel.text    = data.price
    }
}
